woocommerce/woocommerce
# Product attributes lookup table creation and filling

### All Submissions:


### Changes proposed in this Pull Request:

This pull request contains the code needed to create the `wp_wc_product_attributes_lookup` and to fill it with data from the existing products in the database.

The table creation and filling is not performed in a traditional data migration; instead, it must be triggered manually via a debug tool (in the _WooCommerce - Status - Tools_ page). This is necessary for incremental development of the full feature: this pull request creates and fills the table, another one will implement live synchronization of the table as products are added and modified, and yet another one will implement the actual usage of the table for filtering products by attribute. Once everything is in place and well tested a one-liner data migration (`wc_get_container()->get(DataRegenerator::class)->initiate_regeneration();`) can be added; meanwhile, the debug tool can be used for testing and debugging.

The table filling procedure is done in steps, just a few products at a time, using Action Scheduler as in the case of regular data migrations; the difference is that all the step scheduling is done by the `DataRegenerator` class itself and not by the data migrations engine.

The debug tools page entries will show only if the feature is enabled. It's disabled by default, to enable it the following code needs to be executed:

```php
wc_get_container()
->get(\Automattic\WooCommerce\Internal\ProductAttributesLookup\LookupDataStore::class)
->show_feature();
```

#### Improvements in the debug tools page

This pull request implements also a few improvements in the _WooCommerce - Status - Tools_ page:

1. The tool definition array entries (defined in `WC_REST_System_Status_Tools_V2_Controller::get_tools` and in the `woocommerce_debug_tools` filter) get two new optional keys:
    * `disabled`: shows the tool button disabled.
    * `needs_refresh`: when running a tool, by default the tool definitions are retrieved first, then the selected tool is executed, then the definitions previously retrieved are rendered. When this key is defined with a value of `true` the tool definitions are retrieved again after execution, this is useful for cases where the tool description or button enable/disable state changes after the tool execution.

2. If a tool execution throws an exception a notice will be shown with the execption message, previously the exception was unhandled.

Closes #29608.

### How to test the changes in this Pull Request:

1. Go to the _WooCommerce - Status - Tools_ page and see that there's nothing new.

2. Add the following code somewhere, e.g. at the end of `woocommerce.php`:

```php
add_action('woocommerce_init', function() {
    wc_get_container()->get(\Automattic\WooCommerce\Internal\ProductAttributesLookup\LookupDataStore::class)->show_feature();
});
```

3. Reload the _WooCommerce - Status - Tools_ page and find a "Create and fill product attributes lookup table" entry.

![image](https://user-images.githubusercontent.com/937723/116562476-794cbb80-a903-11eb-9f77-1acb0a788793.png)

4. Click the "Create" button. After the page reloads you'll see that the button is disabled and shows a number corresponding to the number of products processed so far (this will also be the case if you reaload the page before the filling process finishes):

![image](https://user-images.githubusercontent.com/937723/116562758-bc0e9380-a903-11eb-8663-ba1f07e1c28b.png)

5. Once the table is completely filled you'll see that there are now three entries: one to start over and regenerate the table data, another one to delete the table, and yet another one to enable the actual usage of the table for filtering:

![image](https://user-images.githubusercontent.com/937723/116563175-13146880-a904-11eb-8ada-0bf0a8848cf4.png)

Of course, "enabling" has no practical effect for now (it just changes the value of the `woocommerce_attribute_lookup__enabled` option). And also of course, when the table usage is enabled the corresponding entry in the tools page becomes a "Disable" one.

6. Verify that the `wp_wc_product_attributes_lookup` table has been created and that it contains the appropriate data. Entries should have been created for all the existing products as per the following rules:
  
    * For simple products: one entry is created per combination of product, taxonomy and term, for all the taxonomies used as product attributes and for all the terms used from that taxonomy.
        * _e.g. a product having attribute A with 4 values and attribute B with 3 values gets 7 entries in the table._
    * For variable products: the same as for simple products, but only for the attributes NOT used for variations.
    * For variation products: One entry for each of the attributes that define the variation, with the appropriate taxonomy – term id pair. **But** if a variation doesn’t have any value for a given attribute (showing then as “Any…” in the UI), then it gets one entry for each possible attribute value.
        * _e.g. a product defines 3 colors and a few sizes for variations. A variation is defined as “size big – any color”. Then that variation gets 4 entries in the table, one for “pa_size = big” and one for each color._
     * Note that custom product attributes don't get entries in the lookup table (they aren't currently available for filtering either).

It may be useful to take a look at [the proof of concept pull request](https://github.com/woocommerce/woocommerce/pull/29190) that contains a hardcoded list of products and a hardcoded SQL statement that creates the corresponding entries in the lookup table.

#### Troubleshooting

If after starting the "Create and fill product attributes lookup table" tool you see its button stuck in "Filling in process" forever that means that something went wrong. To verify, go to _WooCommerce - Tools - Scheduled Actions_ and filter by an action named `woocommerce_run_product_attribute_lookup_update_callback`; the most recent one is likely to be in "Failed" state.

In order to perform local/step-by-step debugging it may be helpful to bypass Action Scheduler and run the entire filling in one single step. You can do that by temporarily replacing the `DataRegenerator::enqueue_regeneration_step_run` method as follows:

```php
private function enqueue_regeneration_step_run() {
  $this->run_regeneration_step_callback();
}
```

You can always go back to a clean state by running the "Delete the product attributes lookup table" tool, even if there's a filling stuck or still in progress. Or from command line:

```
wp eval "wc_get_container()->get(Automattic\WooCommerce\Internal\ProductAttributesLookup\DataRegenerator::class)->delete_all_attributes_lookup_data();"
```

### Other information:


### Changelog entry

> Add - Product attributes lookup table and debug tools to manually fill or delete it.
